content/en/docs/contributing/style-guide.md

```diff
@@ -0,0 +1,73 @@
+---
+title: "Style Guide"
+linkTitle: "Style Guide"
+weight: 80
+type: "docs"
+---
+
+cert-manager, like most Go projects, delegates almost all stylistic choices to `gofmt`,
+with `goimports` on top for organizing imports. Broadly speaking, if you set your editor to run
+`goimports` when you save a file, your code will be stylistically correct.
+
+cert-manager generally also follows the Kubernetes
+[coding conventions](https://www.kubernetes.dev/docs/guide/coding-convention/) and the Google
+[Go code review comments](https://github.com/golang/go/wiki/CodeReviewComments).
+
+## Organizing Imports
+
+Imports should be organized into 3 blocks, with each block separated by two newlines:
+
+```go
+import (
+	"stdlib"
+
+	"external"
+
+	"internal"
+)
+```
+
+An example might be the following, taken from
+[`pkg/acme/accounts/client.go`](https://github.com/jetstack/cert-manager/blob/0c71fe7795858b96cabcddabf706d997cd2fba3f/pkg/acme/accounts/client.go):
+
+```go
+
+import (
+	"crypto/rsa"
+	"crypto/tls"
+	"net"
+	"net/http"
+	"time"
+
+	acmeapi "golang.org/x/crypto/acme"
+
+	acmecl "github.com/jetstack/cert-manager/pkg/acme/client"
+	acmeutil "github.com/jetstack/cert-manager/pkg/acme/util"
+	cmacme "github.com/jetstack/cert-manager/pkg/apis/acme/v1"
+	"github.com/jetstack/cert-manager/pkg/metrics"
+	"github.com/jetstack/cert-manager/pkg/util"
+)
+```
+
+Once this manual split of standard library, external and internal imports has been made, it will be
+enforced automatically by `goimports` when executed in the future.
+
+## Go Module Compatibility
+
+It's important to note that cert-manager `v1` doesn't generally provide any guarantees that it can be
+imported and used as a module by other programs, even for publicly exported functions, variables,
+constants, interfaces and types under `pkg/`.
+
+Only the following modules can be explicitly relied upon to maintain the go module compatibility
+guarantee in cert-manager `v1`:
+
+- `pkg/api`
+- `pkg/apis`
+- `pkg/acme/webhook`
+- `pkg/client`
```

Comment:
This should not be in this list - consumers of our repository should generate their own typed clients rather than depend on ours, as they can (and have) changed between k8s releases that we depend upon

Reply:
Do we count as consumers of cert-manager? :grin:  https://github.com/cert-manager/istio-csr/blob/master/pkg/server/server.go#L29

Reply:
Yes we do - however given we control all these things, it's far more manageable for us to manage these updates in tandem. It's a well known issue when depending on externally generated k8s clients whereby you can get yourself into impossible dependency situations 😬

Reply:
I'd appreciate a link with details about the well known issue because it wasn't well known to me as a relative k8s beginner :grin:

In any case, I'll remove `pkg/client`

Reply:
I can try and find a link, but it typically manifests itself because:

1) code-generator does not guarantee Go API compatible output between minor releases of k8s
2) the things that are generated by code-generator can also be sensitive to the version of apimachinery being used between minor versions
3) this means between two minor versions of our k8s.io dependencies, the generated clients/packages here can change in non Go API compatible ways.

I know we have run into this in the past (hence why we have also been careful to not depend on too many external projects that also depend on k8s things - notable ones that we do depend on are controller-runtime, and for this one I have had to make PRs to bump their k8s dependency versions so that we are able to upgrade _numerous times_)

Reply:
That's super valuable, thanks very much for taking the time to explain, it's much appreciated :+1: